src/io/frame.rs

```diff
@@ -216,57 +222,94 @@ pub trait Decode: Sized {
     /// Finally, if the bytes in the buffer are malformed then an error is
     /// returned indicating why. This informs `Framed` that the stream is now
     /// corrupt and should be terminated.
-    fn decode(buf: &mut EasyBuf) -> Result<Option<Self>, io::Error>;
+    fn decode(&mut self, buf: &mut EasyBuf) -> Result<Option<Self::In>, io::Error>;
 
     /// A default method available to be called when there are no more bytes
     /// available to be read from the underlying I/O.
     ///
     /// This method defaults to calling `decode` and returns an error if
     /// `Ok(None)` is returned. Typically this doesn't need to be implemented
     /// unless the framing protocol differs near the end of the stream.
-    fn done(buf: &mut EasyBuf) -> io::Result<Self> {
-        match try!(Self::decode(buf)) {
+    fn decode_eof(&mut self, buf: &mut EasyBuf) -> io::Result<Self::In> {
+        match try!(self.decode(buf)) {
             Some(frame) => Ok(frame),
             None => Err(io::Error::new(io::ErrorKind::Other,
                                        "bytes remaining on stream")),
         }
     }
-}
 
-/// A trait for encoding frames into a byte buffer.
-///
-/// This trait is used as a building block of `Framed` to define how frames are
-/// encoded into bytes to get passed to the underlying byte stream. Each
-/// frame written to `Framed` will be encoded with this trait to an internal
-/// buffer. That buffer is then written out when possible to the underlying I/O
-/// stream.
-pub trait Encode {
     /// Encodes a frame into the buffer provided.
     ///
     /// This method will encode `msg` into the byte buffer provided by `buf`.
     /// The `buf` provided is an internal buffer of the `Framed` instance and
     /// will be written out when possible.
-    fn encode(self, buf: &mut Vec<u8>);
+    fn encode(&mut self, msg: Self::Out, buf: &mut Vec<u8>);
 }
 
-struct ReadState {
-    eof: bool,
-    is_readable: bool,
-    rd: EasyBuf,
+/// A `Stream` interface to an underlying `Io` object, using the `Decode` trait
+/// to decode frames.
+pub struct FramedRead<T, C> {
+    framed: BiLock<Framed<T, C>>,
 }
 
-impl ReadState {
-    fn new() -> ReadState {
-        ReadState {
-            eof: false,
-            is_readable: false,
-            rd: EasyBuf::new(),
+impl<T: Io, C: Codec> Stream for FramedRead<T, C> {
+    type Item = C::In;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<Option<C::In>, io::Error> {
+        if let Async::Ready(mut guard) = self.framed.poll_lock() {
```

Comment:
This is where I *slightly* regret `if let`, the `match` clause I feel is more concise here:

```rust
match self.framed.poll_lock() {
    Async::Ready(mut guard) => guard.poll(),
    Async::NotReady => Ok(Async::NotReady),
}
```

Just interesting trivia!